File: lib/estimate.dart
import 'package:hands_on_rl/bandit.dart';

List<double> estimate(
  Bandit bandit,
  int Function(int, int) strategy,
  int steps,
) {
  int action = 1;
  int reward = bandit(action);
  final regrets = <double>[];
  for (var t = 0; t < steps; t++) {
    action = strategy(reward, action);
    reward = bandit(action);
    regrets.add(bandit.bestQ - bandit.qs[action]);
  }
  return regrets;
}

File: python/hands_on_rl/estimate.py
from typing import List, Callable
from .bandit import Bandit
import numpy as np


def estimate(
    bandit: Bandit, strategy: Callable[[int, int], int], steps: int
) -> List[float]:
    action = np.random.randint(0, bandit.n - 1)
    reward = bandit(action)
    regrets = []

    for t in range(steps):
        action = strategy(reward, action)
        reward = bandit(action)
        regret = bandit.best_q - bandit.qs[action]
        regrets.append(regret)

    return regrets
